src/cmd/ksh93/tests/alias.sh

```diff
@@ -104,5 +104,172 @@ unalias foo && err_exit 'unalias should return non-zero when a previously set al
 err=$(set +x; { "$SHELL" -i -c 'unalias history; unalias r'; } 2>&1) && [[ -z $err ]] \
 || err_exit "the 'history' and 'r' aliases can't be removed (got $(printf %q "$err"))"
 
+# ======
+# Listing aliases in a script shouldn't break shcomp bytecode
+exp="alias foo='bar('
+alias four=4
+alias three=3
+alias two=2
+foo='bar('
+foo='bar('
+four=4
+three=3
+two=2
+foo='bar('
+four=4
+three=3
+two=2
+noalias: alias not found
+ls=$(whence -p ls)
+alias -t ls"
+alias_script=$tmp/alias_script.sh
+cat > "$alias_script" << EOF
+unalias -a
+alias foo='bar('
+alias two=2
+alias three=3
+alias four=4
+alias -p
+alias foo
+alias -x
+alias
+alias noalias
+alias -t ls
+alias -t
+alias -pt
+EOF
+got=$(redirect 2>&1; $SHELL <($SHCOMP "$alias_script"))
+[[ $exp == $got ]] || err_exit "Listing aliases corrupts shcomp bytecode" \
+	"(expected $(printf %q "$exp"), got $(printf %q "$got"))"
+
+# Specifying an alias with 'alias -p' should print that alias in a reusable form
+exp='alias foo=BAR'
+got=$(
+	alias foo=BAR bar=FOO
+	alias -p foo
+)
+[[ $exp == $got ]] || err_exit "Specifying an alias with 'alias -p' doesn't work" \
+	"(expected $(printf %q "$exp"), got $(printf %q "$got"))"
+
+exp="cat=$(whence -p cat)
+alias -t cat"
+got=$(
+	hash -r cat
+	alias -tx
+	alias -ptx
```

Comment:
Actually, due to the `-x` option, neither of those two alias commands should produce any output. The fact that they do is a bug I introduced in 80d9ae2b1ce0550877bbf394bfb09b186df3cdac. The fix is to restore this flag bit when using `-x`:
```diff
--- a/src/cmd/ksh93/bltins/typeset.c
+++ b/src/cmd/ksh93/bltins/typeset.c
@@ -158,6 +158,7 @@ int    b_alias(int argc,register char *argv[],Shbltin_t *context)
 			break;
 		    case 'x':
 			/* obsolete, ignored */
+			flag |= NV_EXPORT;
 			break;
 		    case 'r':
 			rflag=1;
```
The documentation should probably be updated to say that `-x` suppresses output and is otherwise ignored.

Reply:
Your patch does work in preventing `alias -ptx` from producing output, although it does introduce a bug:
```sh
$ arch/*/bin/ksh -c 'alias foo=bar; alias -x foo; alias -px foo'
alias -x alias foo=bar
```
Here's an improved patch:
```diff
--- a/src/cmd/ksh93/bltins/typeset.c
+++ b/src/cmd/ksh93/bltins/typeset.c
@@ -158,6 +158,7 @@ int    b_alias(int argc,register char *argv[],Shbltin_t *context)
 			break;
 		    case 'x':
 			/* obsolete, ignored */
+			flag |= NV_EXPORT;
 			break;
 		    case 'r':
 			rflag=1;
@@ -605,7 +606,7 @@ static void print_value(Sfio_t *iop, Namval_t *np, struct tdata *tp)
 		sfwrite(iop,"}\n",2);
 		return;
 	}
-	if(tp->prefix && *tp->prefix=='a' && !nv_isattr(np,NV_TAGGED))
+	if(tp->prefix && *tp->prefix=='a' && !nv_isattr(np,NV_TAGGED|NV_EXPORT))
 		sfprintf(iop,"%s ", tp->prefix);
 	table = tp->sh->last_table;
 	sfputr(iop,nv_name(np),aflag=='+'?'\n':'=');
```
This patch causes ksh to treat aliases marked with `-x` differently from normal aliases, since aliases created or modified by the `-x` option are now marked with `NV_EXPORT` (ksh93u+ behaves identically):
```sh
$ arch/*/bin/ksh -c 'alias foo=bar bar=foo; alias -x foo; alias -px'
alias foo=bar
```
Does this LGTY?

Reply:
The patch I posted above has a few inconsistencies. Try this one instead:

<details open>

<summary>alias-x.patch</summary>

```diff
diff --git a/src/cmd/ksh93/bltins/typeset.c b/src/cmd/ksh93/bltins/typeset.c
index db40546a..ba0ffcdd 100644
--- a/src/cmd/ksh93/bltins/typeset.c
+++ b/src/cmd/ksh93/bltins/typeset.c
@@ -132,7 +132,7 @@ int    b_alias(int argc,register char *argv[],Shbltin_t *context)
 {
 	register unsigned flag = NV_NOARRAY|NV_NOSCOPE|NV_ASSIGN;
 	register Dt_t *troot;
-	register int rflag=0, n;
+	register int rflag=0, xflag=0, n;
 	struct tdata tdata;
 	NOT_USED(argc);
 	memset((void*)&tdata,0,sizeof(tdata));
@@ -158,6 +158,8 @@ int    b_alias(int argc,register char *argv[],Shbltin_t *context)
 			break;
 		    case 'x':
 			/* obsolete, ignored */
+			xflag = 1;
+			flag |= NV_EXPORT;
 			break;
 		    case 'r':
 			rflag=1;
@@ -193,6 +195,8 @@ int    b_alias(int argc,register char *argv[],Shbltin_t *context)
 	}
 	else if(argv[1] && tdata.sh->subshell && !tdata.sh->subshare)
 		sh_subfork();			/* avoid affecting the parent shell's alias table */
+	if(xflag && tdata.pflag && (flag&NV_TAGGED))
+		return(0);  /* do nothing */
 	return(setall(argv,flag,troot,&tdata));
 }
 
@@ -605,7 +609,7 @@ static void print_value(Sfio_t *iop, Namval_t *np, struct tdata *tp)
 		sfwrite(iop,"}\n",2);
 		return;
 	}
-	if(tp->prefix && *tp->prefix=='a' && !nv_isattr(np,NV_TAGGED))
+	if(tp->prefix && *tp->prefix=='a' && !nv_isattr(np,NV_TAGGED|NV_EXPORT))
 		sfprintf(iop,"%s ", tp->prefix);
 	table = tp->sh->last_table;
 	sfputr(iop,nv_name(np),aflag=='+'?'\n':'=');
@@ -1448,7 +1452,7 @@ static int print_namval(Sfio_t *file,register Namval_t *np,register int flag, st
 	if(tp->prefix)
 	{
 		outname = (*tp->prefix=='t' && (!nv_isnull(np) || nv_isattr(np,NV_FLOAT|NV_RDONLY|NV_BINARY|NV_RJUST|NV_NOPRINT)));
-		if(tp->scanroot==sh.track_tree && *tp->prefix=='a')
+		if(*tp->prefix=='a' && tp->scanroot==sh.track_tree)
 			tp->prefix = "alias -t";
 		if(indent && (isfun || outname || *tp->prefix!='t'))
 		{
@@ -1461,6 +1465,8 @@ static int print_namval(Sfio_t *file,register Namval_t *np,register int flag, st
 				nv_attribute(np,tp->outfile,tp->prefix,tp->aflag);
 			else
 				sfputr(file,tp->prefix,' ');
+			if(tp->scanroot==sh.alias_tree && nv_isattr(np,NV_EXPORT))
+				sfprintf(file,"-x ");
 		}
 	}
 	if(isfun)
diff --git a/src/cmd/ksh93/tests/alias.sh b/src/cmd/ksh93/tests/alias.sh
index b75d0c4a..7892308f 100755
--- a/src/cmd/ksh93/tests/alias.sh
+++ b/src/cmd/ksh93/tests/alias.sh
@@ -115,23 +115,15 @@ foo='bar('
 four=4
 three=3
 two=2
-foo='bar('
-four=4
-three=3
-two=2
 noalias: alias not found
 ls=$(whence -p ls)
 alias -t ls"
 alias_script=$tmp/alias_script.sh
 cat > "$alias_script" << EOF
 unalias -a
-alias foo='bar('
-alias two=2
-alias three=3
-alias four=4
+alias foo='bar(' two=2 three=3 four=4
 alias -p
 alias foo
-alias -x
 alias
 alias noalias
 alias -t ls
@@ -151,15 +143,19 @@ got=$(
 [[ $exp == $got ]] || err_exit "Specifying an alias with 'alias -p' doesn't work" \
 	"(expected $(printf %q "$exp"), got $(printf %q "$got"))"
 
-exp="cat=$(whence -p cat)
-alias -t cat"
 got=$(
+	redirect 2>&1
+	unalias -a
 	hash -r cat
 	alias -tx
 	alias -ptx
+	alias -ptx cat
+	hash -r
+	alias -ptx chmod
+	hash
 )
-[[ $exp == $got ]] || err_exit "The alias command doesn't ignore the obsolete -x option when -t is passed" \
-	"(expected $(printf %q "$exp"), got $(printf %q "$got"))"
+[[ -z $got ]] || err_exit "'alias -tx' and 'alias -ptx' shouldn't produce output" \
+	"(got $(printf %q "$got"))"
 
 # Listing members of the hash table with 'alias -pt' should work
 exp='alias -t cat
```

</details>

Reply:
Actually, just ignore the above patches. I've opted to make `alias -tx` and `alias -ptx` no-op commands instead (the documentation has also been updated). This was done without setting `NV_EXPORT`, so aliases created with `alias -x` aren't any different from a normal alias.